tailcfg/tailcfg.go

```diff
@@ -694,6 +695,17 @@ type FilterRule struct {
 	// DstPorts are the port ranges to allow once a source IP
 	// matches (is in the CIDR described by SrcIPs & SrcBits).
 	DstPorts []NetPortRange
+
+	// IPProto is the IP protocol number to match.
+	// As a special case, zero matches both TCP (6) and UDP (17).
+	//
+	// Numbers outside the uint8 range (below 0 or above 255) are
+	// reserved for Tailscale's use.
+	//
+	// To actually specify zero, use 256.
+	//
+	// Depending on the IPProto, DstPorts may or may not be used.
+	IPProto int `json:",omitempty"`
```

Comment:
Maybe a slice? Allows for more compact encoding of "any protocol with a port number", i.e. our current behavior. Also gives legacy behavior a very distinct encoding, vs. "special interpretation of zero as not-a-protocol".

Not a strong opinion, just a thought.

Reply:
slice works. thoughts on special representation of all IPProtos? -1? explicitly list all 256?

if we do slice, I guess the PortRange can only apply to some of the protos and be ignored for others? But that's kinda how it is now.

Reply:
I think internally using numbers above 256 for special meanings seems ok but with short string aliases for a better experience?

Alternate implementation: it could be a 256 wide bit field instead of needing to deal with slices.

Reply:
> Alternate implementation: it could be a 256 wide bit field instead of needing to deal with slices.

While cute, it doesn't fit the spirit of the existing JSON verbose style.

We can compile the execution engine down to something faster once we get the JSON once. (we only send it on changes)

Reply:
Made it a slice.

Reply:
Oh cool, thanks github for eating my long comment when I hit unresolve. Gah.

"ports are only applied for protocols that have ports" is fine, that's a common pattern in network-y data structures.

Representing all protos: I think I'd rather avoid having such a representation, and in this protocol layer only support explicitly listing protocols we've thought through and decided to make work (especially since the filtering engine in the client will drop any protocol _it_ doesn't recognize, so a wildcard here is a bit misleading). The ACL compiler in control can support wildcard protocols, and expand them to the specific list. Thoughts?